notebooks/extraction_module/ner_exploration/displaCy_ipywidgets.ipynb

```diff
@@ -0,0 +1,270 @@
+    "                hue, saturation, lightness\n",
+    "            )\n",
+    "        )\n",
+    "    return colours\n",
+    "\n",
+    "\n",
+    "def highlight_character_ranges(text, entities):\n",
+    "    # Initialize an empty list to store formatted HTML for each character\n",
+    "    formatted_chars = []\n",
+    "    # Create a dictionary to map character positions to their corresponding labels\n",
+    "    position_to_label = {}\n",
+    "    colour_of_label = {}\n",
+    "\n",
+    "    labels = [e[\"label\"] for e in entities]\n",
+    "    start_ids = [e[\"start\"] for e in entities]\n",
+    "    end_ids = [e[\"end\"] for e in entities]\n",
+    "\n",
+    "    colours = generate_colours(len(labels), transparency=0.5)\n",
+    "\n",
+    "    # Iterate through each range of character positions and labels\n",
+    "    for label, start_idx, end_idx, colour in zip(\n",
+    "        labels, start_ids, end_ids, colours\n",
+    "    ):\n",
+    "        # Store the label for each character position within the range\n",
+    "        for i in range(start_idx, end_idx + 1):\n",
+    "            position_to_label[i] = label\n",
+    "            colour_of_label[i] = colour\n",
+    "\n",
+    "    # Iterate through each character in the text string\n",
+    "    for i, char in enumerate(text):\n",
+    "        # Check if the current character position should be highlighted\n",
+    "        if i in position_to_label:\n",
+    "            # If the character position should be highlighted, wrap it in a span with CSS class for styling and tooltip\n",
+    "            formatted_chars.append(\n",
+    "                f'<span class=\"highlight\" style=\"background-color: {colour_of_label[i]};\" title=\"{position_to_label[i]}\">{text[i]}</span>'\n",
+    "            )\n",
+    "        else:\n",
+    "            # If the character position shouldn't be highlighted, just use the character as is\n",
+    "            formatted_chars.append(char)\n",
+    "\n",
+    "    # Join the formatted characters back into a single string\n",
+    "    highlighted_text = \"\".join(formatted_chars)\n",
+    "\n",
+    "    # Create custom CSS style for the highlight class\n",
+    "    custom_css = \"\"\"\n",
+    "    <style>\n",
+    "    .highlight {\n",
+    "        cursor: pointer; /* Change cursor to pointer on hover */\n",
+    "    }\n",
+    "    </style>\n",
+    "    \"\"\"\n",
+    "\n",
+    "    # Create an HTML widget to display the highlighted text\n",
+    "    html_widget = widgets.HTML(value=custom_css + highlighted_text)\n",
+    "\n",
+    "    # Display the HTML widget\n",
+    "    display(html_widget, clear=True)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "text = string_data[string_id]\n",
+    "entities = entity_data[string_id][\"Entities\"]\n",
+    "\n",
+    "highlight_character_ranges(text, entities)"
```

Comment:
Might be good to print the entity names after the output with them also coloured as a key as a future extension?